In [16]:
#импорт библиотек
import torch            #тензорные вычисления
import numpy as np      #массивы и матрицы и математика
import pandas as pd     #инструмент анализа и манипулирования c данными
import sklearn as sk    #бибоиотека алгоритмов машинного обучения
import seaborn as sns   #визуализация

Машинное обучение
Задача регрессии

Предсказание цены на подеражанные автомобили
на датасете: https://www.kaggle.com/datasets/thedevastator/uncovering-factors-that-affect-used-car-prices

In [17]:
#загрузка датасета
#датасет объявлений с продажей подержанных автомобилей
df=pd.read_csv('autos.csv')

In [18]:
#посмотрим на структуру датасета
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 21 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   index                371528 non-null  int64 
 1   dateCrawled          371528 non-null  object
 2   name                 371528 non-null  object
 3   seller               371528 non-null  object
 4   offerType            371528 non-null  object
 5   price                371528 non-null  int64 
 6   abtest               371528 non-null  object
 7   vehicleType          333659 non-null  object
 8   yearOfRegistration   371528 non-null  int64 
 9   gearbox              351319 non-null  object
 10  powerPS              371528 non-null  int64 
 11  model                351044 non-null  object
 12  kilometer            371528 non-null  int64 
 13  monthOfRegistration  371528 non-null  int64 
 14  fuelType             338142 non-null  object
 15  brand                371528 non-nu

In [19]:
#посмотрим на данные
df.head(15)

index          dateCrawled  \
0       0  2016-03-24 11:52:17   
1       1  2016-03-24 10:58:45   
2       2  2016-03-14 12:52:21   
3       3  2016-03-17 16:54:04   
4       4  2016-03-31 17:25:20   
5       5  2016-04-04 17:36:23   
6       6  2016-04-01 20:48:51   
7       7  2016-03-21 18:54:38   
8       8  2016-04-04 23:42:13   
9       9  2016-03-17 10:53:50   
10     10  2016-03-26 19:54:18   
11     11  2016-04-07 10:06:22   
12     12  2016-03-15 22:49:09   
13     13  2016-03-21 21:37:40   
14     14  2016-03-21 12:57:01   

                                                 name  seller offerType  \
0                                          Golf_3_1.6  privat   Angebot   
1                                A5_Sportback_2.7_Tdi  privat   Angebot   
2                      Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3                                  GOLF_4_1_4__3TÜRER  privat   Angebot   
4                      Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   
5   BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...  privat   Angebot   
6                         Peugeot_206_CC_110_Platinum  privat   Angebot   
7                        VW_Derby_Bj_80__Scheunenfund  privat   Angebot   
8                Ford_C___Max_Titanium_1_0_L_EcoBoost  privat   Angebot   
9   VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...  privat   Angebot   
10                                  Mazda_3_1.6_Sport  privat   Angebot   
11      Volkswagen_Passat_Variant_2.0_TDI_Comfortline  privat   Angebot   
12                  VW_Passat_Facelift_35i__"7Sitzer"  privat   Angebot   
13                     VW_PASSAT_1.9_TDI_131_PS_LEDER  privat   Angebot   
14  Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...  privat   Angebot   

    price   abtest vehicleType  yearOfRegistration    gearbox  ...    model  \
0     480     test         NaN                1993    manuell  ...     golf   
1   18300     test       coupe                2011    manuell  ...      NaN   
2    9800     test         suv                2004  automatik  ...    grand   
3    1500     test  kleinwagen                2001    manuell  ...     golf   
4    3600     test  kleinwagen                2008    manuell  ...    fabia   
5     650     test   limousine                1995    manuell  ...      3er   
6    2200     test      cabrio                2004    manuell  ...  2_reihe   
7       0     test   limousine                1980    manuell  ...   andere   
8   14500  control         bus                2014    manuell  ...    c_max   
9     999     test  kleinwagen                1998    manuell  ...     golf   
10   2000  control   limousine                2004    manuell  ...  3_reihe   
11   2799  control       kombi                2005    manuell  ...   passat   
12    999  control       kombi                1995    manuell  ...   passat   
13   2500  control       kombi                2004    manuell  ...   passat   
14  17999  control         suv                2011    manuell  ...   navara   

   kilometer  monthOfRegistration  fuelType       brand notRepairedDamage  \
0     150000                    0    benzin  volkswagen               NaN   
1     125000                    5    diesel        audi                ja   
2     125000                    8    diesel        jeep               NaN   
3     150000                    6    benzin  volkswagen              nein   
4      90000                    7    diesel       skoda              nein   
5     150000                   10    benzin         bmw                ja   
6     150000                    8    benzin     peugeot              nein   
7      40000                    7    benzin  volkswagen              nein   
8      30000                    8    benzin        ford               NaN   
9     150000                    0       NaN  volkswagen               NaN   
10    150000                   12    benzin       mazda              nein   
11    150000                   12    diesel  volkswagen              

In [20]:
df

index          dateCrawled  \
0            0  2016-03-24 11:52:17   
1            1  2016-03-24 10:58:45   
2            2  2016-03-14 12:52:21   
3            3  2016-03-17 16:54:04   
4            4  2016-03-31 17:25:20   
...        ...                  ...   
371523  371523  2016-03-14 17:48:27   
371524  371524  2016-03-05 19:56:21   
371525  371525  2016-03-19 18:57:12   
371526  371526  2016-03-20 19:41:08   
371527  371527  2016-03-07 19:39:19   

                                                name  seller offerType  price  \
0                                         Golf_3_1.6  privat   Angebot    480   
1                               A5_Sportback_2.7_Tdi  privat   Angebot  18300   
2                     Jeep_Grand_Cherokee_"Overland"  privat   Angebot   9800   
3                                 GOLF_4_1_4__3TÜRER  privat   Angebot   1500   
4                     Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   3600   
...                                              ...     ...       ...    ...   
371523                    Suche_t4___vito_ab_6_sitze  privat   Angebot   2200   
371524         Smart_smart_leistungssteigerung_100ps  privat   Angebot   1199   
371525            Volkswagen_Multivan_T4_TDI_7DC_UY2  privat   Angebot   9200   
371526                        VW_Golf_Kombi_1_9l_TDI  privat   Angebot   3400   
371527  BMW_M135i_vollausgestattet_NP_52.720____Euro  privat   Angebot  28990   

         abtest vehicleType  yearOfRegistration    gearbox  ...        model  \
0          test         NaN                1993    manuell  ...         golf   
1          test       coupe                2011    manuell  ...          NaN   
2          test         suv                2004  automatik  ...        grand   
3          test  kleinwagen                2001    manuell  ...         golf   
4          test  kleinwagen                2008    manuell  ...        fabia   
...         ...         ...                 ...        ...  ...          ...   
371523     test         NaN                2005        NaN  ...          NaN   
371524     test      cabrio                2000  automatik  ...       fortwo   
371525     test         bus                1996    manuell  ...  transporter   
371526     test       kombi                2002    manuell  ...         golf   
371527  control   limousine                2013    manuell  ...      m_reihe   

       kilometer  monthOfRegistration  fuelType           brand  \
0         150000                    0    benzin      volkswagen   
1         125000                    5    diesel            audi   
2         125000                    8    diesel            jeep   
3         150000                    6    benzin      volkswagen   
4          90000                    7    diesel           skoda   
...          ...                  ...       ...             ...   
371523     20000                    1       NaN  sonstige_autos   
371524    125000                    3    benzin           smart   
371525    150000                    3    diesel      volkswagen   
371526    150000                    6    diesel      volkswagen   
371527     50000                    8    benzin             bmw   

       notRepairedDamage          dateCreated nrOfPictures  postalCode  \
0                    NaN  2016-03-24 00:00:00            0       70435   
1                     ja  2016-03-24 00:00:00            0       66954   
2                    NaN  2016-03-14 00:00:00            0       90480   
3                   nein  2016-03-17 00:00:00            0       91074   
4                   nein  2016-03-31 00:00:00            0       60437   
...                  ...                  ...          ...         ...   
371523               NaN  2016-03-14 00:00:00            0       39576   
371524              nein  2016-03-05 00:00:00            0       26135   
371525              nein  2016-03-19 00:00:00            0       87439   
371526               NaN  2016-03-20 00:00:00            0   

In [21]:
df.describe(include='all')

index          dateCrawled         name  seller offerType  \
count   371528.000000               371528       371528  371528    371528   
unique            NaN               280500       233531       2         2   
top               NaN  2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot   
freq              NaN                    7          657  371525    371516   
mean    185763.500000                  NaN          NaN     NaN       NaN   
std     107251.039743                  NaN          NaN     NaN       NaN   
min          0.000000                  NaN          NaN     NaN       NaN   
25%      92881.750000                  NaN          NaN     NaN       NaN   
50%     185763.500000                  NaN          NaN     NaN       NaN   
75%     278645.250000                  NaN          NaN     NaN       NaN   
max     371527.000000                  NaN          NaN     NaN       NaN   

               price  abtest vehicleType  yearOfRegistration  gearbox  ...  \
count   3.715280e+05  371528      333659       371528.000000   351319  ...   
unique           NaN       2           8                 NaN        2  ...   
top              NaN    test   limousine                 NaN  manuell  ...   
freq             NaN  192585       95894                 NaN   274214  ...   
mean    1.729514e+04     NaN         NaN         2004.577997      NaN  ...   
std     3.587954e+06     NaN         NaN           92.866598      NaN  ...   
min     0.000000e+00     NaN         NaN         1000.000000      NaN  ...   
25%     1.150000e+03     NaN         NaN         1999.000000      NaN  ...   
50%     2.950000e+03     NaN         NaN         2003.000000      NaN  ...   
75%     7.200000e+03     NaN         NaN         2008.000000      NaN  ...   
max     2.147484e+09     NaN         NaN         9999.000000      NaN  ...   

         model      kilometer  monthOfRegistration  fuelType       brand  \
count   351044  371528.000000        371528.000000    338142      371528   
unique     251            NaN                  NaN         7          40   
top       golf            NaN                  NaN    benzin  volkswagen   
freq     30070            NaN                  NaN    223857       79640   
mean       NaN  125618.688228             5.734445       NaN         NaN   
std        NaN   40112.337051             3.712412       NaN         NaN   
min        NaN    5000.000000             0.000000       NaN         NaN   
25%        NaN  125000.000000             3.000000       NaN         NaN   
50%        NaN  150000.000000             6.000000       NaN         NaN   
75%        NaN  150000.000000             9.000000       NaN         NaN   
max        NaN  150000.000000            12.000000       NaN         NaN   

       notRepairedDamage          dateCreated nrOfPictures    postalCode  \
count             299468               371528     371528.0  371528.00000   
unique                 2                  114          NaN           NaN   
top                 nein  2016-04-03 00:00:00          NaN           NaN   
freq              263182                14450          NaN           NaN   
mean                 NaN                  NaN          0.0   50820.66764   
std                  NaN                  NaN          0.0   25799.08247   
min                  NaN                  NaN          0.0    1067.00000   
25%                  NaN                  NaN          0.0   30459.00000   
50%                  NaN                  NaN          0.0   49610.00000   
75%                  NaN                  NaN          0.0   71546.00000   
max                  NaN                  NaN          0.0   99998.00000   

                   lastSeen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN 

Таблица с данными содержит 371528 строк

Опишем столбцы-признаки:

index           = номер объявления по порядку начиная с 0
dateCrawled     = дата размещения
name            = название автомобиля
seller          = Тип продавца (частное объявление или диллер)
offerType       = Тип предложения
price           = собственно цена (этот столбец сделаем y=правильными ответами для обучения с учителем)
abtest          =
vehicleType     = тип автомобиля (седан, универсал и т.д.)
yearOfRegistration = год регистрации
gearbox         = тип коробки передач (ручная, автоматическая)
powerPS         = мощность автомобиля
model           = модель автомобиля
kilometer       = пробег
monthOfRegistration = месяц когда автомобиль был зарегистрирован
fuelType        = тип топлива (бензин, дизель или иное)
brand           = марка автомобиля
notRepairedDamage = наличие повреждений, которые не были устранены
dateCreated     = дата выпуска авто
nrOfPictures    = количество фотографий автомобиля
postalCode      = почтовый индекс авто (пожалуй можно назвать уточнением региона)
lastSeen        = дата когда, машину было видно последний раз

